src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs

```diff
@@ -98,6 +98,22 @@ public TypeFactory(OutputLibrary library)
             _ => _library.FindTypeForSchema(schema).WithNullable(isNullable)
         };
 
+        private CSharpType CreateTypeFromUnion(InputUnionType unionType)
+        {
+            if (unionType.UnionItemTypes.Count != 2 || !unionType.UnionItemTypes.Any(t => t is InputIntrinsicType { Kind: InputIntrinsicTypeKind.Null }))
+            {
+                return CreateType(unionType.UnionItemTypes[0]); //TODO -- need to support multiple union types.
+            }
+
+            var nullableTypeCandidate = unionType.UnionItemTypes.FirstOrDefault(t => t is InputPrimitiveType or InputModelType);
+            if (nullableTypeCandidate == null)
+            {
+                return CreateType(unionType.UnionItemTypes[0]);
+            }
+
+            return CreateType(nullableTypeCandidate with { IsNullable = true });
+        }
```

Comment:
I kind of disagree this part, we should not treat something like `int | null` as a `union` despite it is in typescript.
In our original plan for unions of concrete types, the union part should throw an exception because dotnet does not have a concept of union, we will never generate a type for InputUnionType directly. Instead, we will do some kind of transformation, which transforms unions to some fundamental things that dotnet could express.
Instead I believe here our DOTNET emitter should do something to hide the fact of `int | null` is a union, instead it should just emit an int with `nullable = true` by producing a `InputPrimitiveType` instead.